doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,190 @@
+
+Doing both of these lets us remove the concept of subsources from Materialize
+entirely. It, by definition, also lets us get rid of the notion of "progress
+subsources," which is a concept that users often find confusion.
+
+## Success Criteria
+
+- Users no longer need to use the keyword `SUBSOURCE` to interact with
+  Materialize in any capacity; and we do this without deprecating any
+  functionality whatsoever.
+
+## Solution Proposal
+
+### SOURCEs should output to TABLES
+
+In the current version of Materialize, we have an impedance mismatch between
+single-output and multi-output sources.
+
+- Single-ouput sources write their data directly to the `SOURCE` itself, i.e.
+  users query the `SOURCE` object to read the data.
+- Multi-output sources write their data to `SUBSOURCES`. The primary source to
+  which they belong is an empty, unused (though still queryable) collection.
+
+I propose that the output of all sources, both single- and multi-output, should
+be to `TABLE`s (albeit special, read-only tables). Users already have an
+intuition around how `TABLE`s work, and understanding "read-only" tables is much
+simpler than doubling down on subsources––a concept we invented.
+
+#### Why not just rename `SUBSOURCE` to `TABLE`?
+
+This leaves the impedance mismatch in place where users have to `SELECT` from
+the source if they're using e.g. Kafka, but have to find the right `TABLE` to
+`SELECT` from for PG sources.
+
+### SOURCES should contain their progress information
+
+The other type of `SUBSOURCE` Materialize supports is known as the
+remap/progress collection. While most users don't access this data, it's an
+important debugging tool.
+
+I propose that the progress collection should be moved "into" the source, i.e.
+querying a source should return the what-is-today-the-progress-collection's
+data.
+
+### New API
+
+The API to support this is similar to the existing `CREATE SOURCE` API. However,
+`ALTER SOURCE...ADD SUBSOURCE` will disappear and be replaced with a flavor of
+`CREATE TABLE`.
+
+(h/t @benesch for this)
+
+***
+
+Create a Kafka source that reads from a topic named `t`:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR ALL TABLES
+```
+
+`kafka_src` is the progress relation for the source, and a table named `t` is
+automatically created in the same schema as `kafka_src`.
+
+Note that we could also probably make `FOR ALL TABLES` implied for Kafka
+sources.
+
+***
+
+Create a Kafka source that reads from a topic named `t` with a custom table
+name:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR TABLES (t AS my_t)
+```
+
+`kafka_src` is again the progress relation for the source. The data is in a
+table named `my_t`. You would never be able to specify more than one table in
+the original `CREATE SOURCE` statement because you would have to two outputs
+from the same source with the same schema, which we should disallow.
```

Comment:
Wait, this seems like a plausibly useful thing to allow? At least, it doesn't seem like something we need to hard ban. I like the idea of just generally allowing people to create as many copies of a table as they like, and if the upstream table changes schemas, the new schemas just naturally get picked up.

Reply:
Sure thing.